File: packages/core/src/roles.ts
/**
 * Access areas — the client half of `api/app/roles.py`.
 *
 * Both copies must agree, and `api/tests/test_access_areas.py` asserts it by
 * reading this file. This copy hides controls a person cannot use; that is a
 * courtesy, not a gate. The server is the only thing that actually refuses.
 *
 * SIX AREAS, HELD IN ANY COMBINATION (V5 §3)
 *
 *   hplProduction  raise tickets, submit production data, read the daily log
 *   maintenance    acknowledge, work and close tickets; the ticket summary
 *   supervisor     notified about flagged tickets; reopen; reassign
 *   manager        the same, plus handing a ticket to a named person
 *   dashboard      the analytics dashboards and their exports
 *   admin          approve accounts, grant areas, edit the master lists
 *
 * A person holds a SET, and their capabilities are the union. Nothing is
 * inherited: an account with `dashboard` alone reads every number and cannot
 * touch a ticket, which is what a plant head usually wants.
 *
 * The previous model was one `role` string of two values. It was right while
 * the only question was "does this person need the board", and wrong the moment
 * V5 asked for a supervisor who is notified about flagged tickets but does not
 * work them.
 */

export const AREAS = [
  'hpl_production',
  'maintenance',
  'supervisor',
  'manager',
  'dashboard',
  'admin',
] as const;
export type Area = (typeof AREAS)[number];

export interface Capabilities {
  raiseTicket: boolean;
  workTicket: boolean;
  reopenTicket: boolean;
  /** Handing a ticket to a named person on somebody else's behalf (V5 §15.5). */
  reassignTicket: boolean;
  logProduction: boolean;
  viewDashboard: boolean;
  editMasters: boolean;
  approveUsers: boolean;
  manageUsers: boolean;
  unlockUsers: boolean;
  exportData: boolean;
  importData: boolean;
}

const NONE: Capabilities = {
  raiseTicket: false,
  workTicket: false,
  reopenTicket: false,
  reassignTicket: false,
  logProduction: false,
  viewDashboard: false,
  editMasters: false,
  approveUsers: false,
  manageUsers: false,
  unlockUsers: false,
  exportData: false,
  importData: false,
};

/**
 * Raising a breakdown is on every area on purpose. A machine that stopped is a
 * fact, not a privilege, and the person standing next to it is whoever happens
 * to be standing next to it. The one account that cannot raise a ticket holds
 * no areas at all — a signup nobody has approved, or one deliberately emptied.
 */
export const CAPABILITIES: Record<Area, Capabilities> = {
  hpl_production: { ...NONE, raiseTicket: true, logProduction: true },
  maintenance: { ...NONE, raiseTicket: true, workTicket: true, reopenTicket: true },
  supervisor: { ...NONE, raiseTicket: true, reopenTicket: true, reassignTicket: true },
  manager: { ...NONE, raiseTicket: true, reopenTicket: true, reassignTicket: true },
  dashboard: { ...NONE, viewDashboard: true, exportData: true },
  admin: {
    ...NONE,
    raiseTicket: true,
    reopenTicket: true,
    reassignTicket: true,
    editMasters: true,
    approveUsers: true,
    manageUsers: true,
    unlockUsers: true,
    importData: true,
  },
};

const KEYS = Object.keys(NONE) as (keyof Capabilities)[];

export function isArea(value: unknown): value is Area {
  return typeof value === 'string' && (AREAS as readonly string[]).includes(value);
}

/** The union over every area held. Unknown names contribute nothing. */
export function capabilitiesOf(areas: readonly string[] | undefined): Capabilities {
  const held = (areas ?? []).filter(isArea).map((a) => CAPABILITIES[a]);
  if (held.length === 0) return NONE;
  return Object.fromEntries(
    KEYS.map((k) => [k, held.some((c) => c[k])]),
  ) as unknown as Capabilities;
}

export function can(
  areas: readonly string[] | undefined,
  capability: keyof Capabilities,
): boolean {
  return capabilitiesOf(areas)[capability];
}

export function seesDashboard(areas: readonly string[] | undefined): boolean {
  return can(areas, 'viewDashboard');
}

/**
 * Whether this account is waiting to be let in.
 *
 * Holding nothing is not enough to tell: an account whose access was revoked
 * also holds nothing, and it has already been dealt with. `approvedAt` is what
 * separates the two.
 */
export function isPendingApproval(user: {
  areas?: readonly string[];
  approved_at?: string | null;
}): boolean {
  return !user.approved_at;
}

/** Where a person lands after signing in. */
export function homeRouteFor(areas: readonly string[] | undefined): string {
  return seesDashboard(areas) ? '/board' : '/floor';
}

File: api/app/roles.py
"""Access areas — the server-side half of `packages/core/src/roles.ts`.

Both copies must agree. This one is the enforcing copy: the TypeScript version
exists to hide controls a person cannot use, which is a courtesy, not a gate.
`tests/test_access_areas.py` asserts the two files stay in step.

SIX AREAS, HELD IN ANY COMBINATION

V5 §3 names them and is explicit that they are independent grants rather than a
ladder. A shift supervisor who also logs production holds two areas, not a
compromise between two roles; there is no cap and no implied ordering. Admin
switches on whichever apply to each person.

    hpl_production  raise tickets, submit production data, read the daily log
    maintenance     acknowledge, work and close tickets; the ticket summary
    supervisor      notified about flagged tickets; reopen; reassign
    manager         the same, plus handing a ticket to a named person
    dashboard       the analytics dashboards and their exports
    admin           approve accounts, grant areas, edit the master lists

A person holds a SET of these, and their capabilities are the union. Nobody
inherits anything: an account with `dashboard` alone can read the board and
cannot touch a ticket, which is exactly what a plant head usually wants.

WHAT THIS REPLACED

`role`, one string of two values — `app` and `dashboard`. It was the right
model while the only access question was "does this person need the board", and
the wrong one the moment V5 asked for a supervisor who is notified about
flagged tickets but does not work them.

The column is DELETED in migration 0014, not left unread. A column nothing
consults still looks like it means something, and the thing it would appear to
mean is the one thing it no longer decides.

WHY REOPEN IS ITS OWN CAPABILITY

It used to travel with closing, under `verify_close`, on the reasoning that
whoever signs a ticket off should not be whoever fixed it. V5 §5.4 settles both
halves differently: every ticket is self-closed by the person who solved it, so
closing is part of working; and reopening is open to Maintenance, Supervisor,
Manager and Admin alike, because the only real safety net against a premature
close is somebody noticing the machine is still broken, and that net is worth
keeping wide.
"""

from dataclasses import dataclass

AREAS: tuple[str, ...] = (
    "hpl_production",
    "maintenance",
    "supervisor",
    "manager",
    "dashboard",
    "admin",
)


@dataclass(frozen=True, slots=True)
class Capabilities:
    """What somebody may do. The union over every area they hold."""

    raise_ticket: bool = False
    work_ticket: bool = False
    reopen_ticket: bool = False
    # Handing a ticket to a named person on somebody else's behalf (V5 §15.5).
    # Distinct from an engineer passing their own ticket on at shift change,
    # which is part of working it.
    reassign_ticket: bool = False
    log_production: bool = False
    view_dashboard: bool = False
    edit_masters: bool = False
    approve_users: bool = False
    manage_users: bool = False
    unlock_users: bool = False
    export_data: bool = False
    import_data: bool = False


# Raising a breakdown is on every area on purpose. A machine that stopped is a
# fact, not a privilege (CLAUDE.md), and the person standing next to it is
# whoever happens to be standing next to it. The one account that cannot raise
# a ticket is one holding no areas at all — a signup nobody has approved, or an
# account whose access was deliberately emptied.
CAPABILITIES: dict[str, Capabilities] = {
    "hpl_production": Capabilities(
        raise_ticket=True,
        log_production=True,
    ),
    "maintenance": Capabilities(
        raise_ticket=True,
        work_ticket=True,
        reopen_ticket=True,
    ),
    "supervisor": Capabilities(
        raise_ticket=True,
        reopen_ticket=True,
        reassign_ticket=True,
    ),
    "manager": Capabilities(
        raise_ticket=True,
        reopen_ticket=True,
        reassign_ticket=True,
    ),
    # Reading the board is not working the floor. Somebody who holds only this
    # can see every number and cannot acknowledge a ticket, which is what a
    # plant head or a visiting director actually needs.
    "dashboard": Capabilities(
        view_dashboard=True,
        export_data=True,
    ),
    "admin": Capabilities(
        raise_ticket=True,
        reopen_ticket=True,
        reassign_ticket=True,
        edit_masters=True,
        approve_users=True,
        manage_users=True,
        unlock_users=True,
        import_data=True,
    ),
}

_FIELDS = tuple(Capabilities.__dataclass_fields__)
NONE = Capabilities()


def capabilities_of(areas: object) -> Capabilities:
    """The union of every area held. Unknown areas contribute nothing.

    An unknown name is ignored rather than raising: the database CHECK is what
    keeps the column honest, and a request that 500s because somebody typed a
    new area into a row is a worse failure than that area simply not granting
    anything yet.
    """
    held = [CAPABILITIES[a] for a in (areas or ()) if a in CAPABILITIES]
    if not held:
        return NONE
    return Capabilities(**{f: any(getattr(c, f) for c in held) for f in _FIELDS})


def can(areas: object, capability: str) -> bool:
    return getattr(capabilities_of(areas), capability, False) is True


def sees_dashboard(areas: object) -> bool:
    return can(areas, "view_dashboard")
